File: source/extensions/Camera/CameraStyles.java
package source.extensions.Camera;

import com.badlogic.gdx.graphics.Camera;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.math.Vector3;

/**
 * Holds a bunch of camera styles.
 *
 * @author Xico
 * @see com.badlogic.gdx.graphics.Camera
 */

public class CameraStyles {

    /**
     * Sets the {@link Camera} position to a new position and updates it.
     *
     * @param camera   camera used.
     * @param position new camera position.
     */
    private static void setAndUpdate(Camera camera, Vector3 position) {
        camera.position.set(position);
        camera.update();
    }

    /**
     * Centers the {@link Camera} in the middle of the screen.
     *
     * @param camera      camera used.
     * @param worldWidth  width of the world (world coordinates).
     * @param worldHeight height of the world (world coordinates).
     */
    public static void centerOnScreen(Camera camera, int worldWidth, int worldHeight) {
        final Vector3 position = camera.position;
        position.x = worldWidth / 2f;
        position.y = worldHeight / 2f;
        setAndUpdate(camera, position);
    }

    /**
     * The {@link Camera} will be 'stuck' to the target, following it rigidly.
     * Not suitable for all cases, as it will cause a rough motion.
     *
     * @param camera         camera used.
     * @param targetPosition position of the target.
     */
    public static void lockOnTarget(Camera camera, Vector2 targetPosition) {
        final Vector3 position = camera.position;
        position.x = targetPosition.x;
        position.y = targetPosition.y;
        setAndUpdate(camera, position);
    }

    /**
     * The {@link Camera} X will be exactly the same as the target's X, but the Y will be offset by a given amount.
     *
     * @param camera         camera used.
     * @param targetPosition position of the target.
     * @param yOffset        camera's Y offset.
     */
    public static void lockOnTargetX(Camera camera, Vector2 targetPosition, float yOffset) {
        final Vector3 position = camera.position;
        position.x = targetPosition.x;
        position.y = targetPosition.y + yOffset;
        setAndUpdate(camera, position);
    }

    /**
     * The {@link Camera} Y will be exactly the same as the target's Y, but the X will be offset by a given amount.
     *
     * @param camera         camera used.
     * @param targetPosition position of the target.
     * @param xOffset        camera's X offset.
     */
    public static void lockOnTargetY(Camera camera, Vector2 targetPosition, float xOffset) {
        final Vector3 position = camera.position;
        position.x = targetPosition.x + xOffset;
        position.y = targetPosition.y;
        setAndUpdate(camera, position);
    }

    /**
     * The {@link Camera} will follow a target with a smooth linear interpolation, based on a lerp amount.
     * The lower the lerp amount, the smoother - and thus slower - the following motion.
     *
     * @param targetPosition position of the target.
     * @param lerp           defines how smoothly the camera follows; slowest -> ]0, 1] <- fastest.
     * @param xOffset        camera's X offset.
     * @param yOffset        camera's Y offset.
     */
    public static void lerpToTarget(Camera camera, Vector2 targetPosition, float lerp, float xOffset, float yOffset) {
        final Vector3 position = camera.position;
        position.x += ((targetPosition.x - position.x) * lerp) + xOffset;
        position.y += ((targetPosition.y - position.y) * lerp) + yOffset;
        setAndUpdate(camera, position);
    }

    /**
     * Overload of the {@link #lerpToTarget(Camera, Vector2, float, float, float)} method.
     */
    public static void lerpToTarget(Camera camera, Vector2 targetPosition, float lerp) {
        lerpToTarget(camera, targetPosition, lerp, 0f, 0f);
    }
}

File: source/extensions/camera/CameraStyles.kt
@file:JvmName("CameraStyles")

package source.extensions.camera

import com.badlogic.gdx.graphics.Camera
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.math.Vector3

/**
 * Centers the [Camera] in the middle of the screen, fiven the world coordinates.
 *
 * @param worldWidth  width of the world (world coordinates).
 * @param worldHeight height of the world (world coordinates).
 */
fun Camera.center(worldWidth: Int, worldHeight: Int) {
  val newPosition: Vector3 = Vector3(worldWidth / 2f, worldHeight / 2f, 0f)
  this.refresh(newPosition)
}

/**
 * The [Camera] will follow the target coordinates rigidly.
 * Not suitable for all cases, as it may cause a very rough motion.
 *
 * @param targetPosition position of the target, where the camera will be also positioned to.
 */
fun Camera.lockOn(targetPosition: Vector2) {
  val newPosition: Vector3 = Vector3(targetPosition.x, targetPosition.y, 0f)
  this.refresh(newPosition)
}

/**
 * The [Camera] will rigidly follow the [targetPosition] horizontally, but will be
 * offset vertically by a [yOffset].
 *
 * @param targetPosition position of the target, where the camera will be also positioned to (horizontally).
 * @param yOffset camera's Y offset, relatively to the [targetPosition] Y.
 */
fun Camera.lockOnX(targetPosition: Vector2, yOffset: Float) {
  val newPosition: Vector3 = Vector3(targetPosition.x, targetPosition.y + yOffset, 0f)
  this.refresh(newPosition)
}

/**
 * The [Camera] will rigidly follow the [targetPosition] vertically, but will be
 * offset horizontally by a [xOffset].
 *
 * @param targetPosition position of the target, where the camera will be also positioned to (horizontally).
 * @param xOffset camera's X offset, relatively to the [targetPosition] X.
 */
fun Camera.lockOnY(targetPosition: Vector2, xOffset: Float) {
  val newPosition: Vector3 = Vector3(targetPosition.x + xOffset, targetPosition.y, 0f)
  this.refresh(newPosition)
}

/**
 * The [Camera] will follow a [targetPosition] with a smooth linear interpolation, based on a [lerp] amount.
 * The lower the lerp amount, the smoother - and thus slower - the following motion.
 * The camera can also be offset by an [xOffset] and [yOffset] amounts, but that's optional.
 *
 * @param targetPosition  position of the target that the camera will be set on.
 * @param lerp            defines how smoothly the camera follows; ]0 (slowest), 1 (fastest)].
 * @param xOffset         camera's X offset, relatively to the [targetPosition] X.
 * @param yOffset         camera's Y offset, relatively to the [targetPosition] Y.
 */
@JvmOverloads
fun Camera.lerpTo(targetPosition: Vector2, lerp: Float, xOffset: Float = 0f, yOffset: Float = 0f) {
  val newX: Float = this.position.x + ((targetPosition.x - position.x) * lerp) + xOffset
  val newY: Float = this.position.y + ((targetPosition.y - position.y) * lerp) + yOffset
  val newPosition: Vector3 = Vector3(newX, newY, 0f)
  this.refresh(newPosition)
}

/**
 * Sets the [Camera] position to the given one and updates it.
 * Won't be called explicitly anywhere besides this class.
 *
 * @param newPosition new [Vector3] position for the camera.
 */
private fun Camera.refresh(newPosition: Vector3) {
  this.apply {
    position.set(newPosition)
    update()
  }
}